### ISSUE TITLE
Support FBX file loading

### ISSUE THREAD

**Issue by tbillington** [bevyengine/bevy#15705]
## What problem does this solve or what need does it fill?

[FBX](https://en.wikipedia.org/wiki/FBX) is a commonly used file format for 3d models. Many 3d models & asset packs from asset stores, eg the popular [Unity AssetStore](https://assetstore.unity.com/), only supply models & materials in `.fbx` format.

By not supporting FBX, Bevy users must convert assets manually which is especially painful as some asset packs can have hundreds of model files. This hinders teams trying to leverage commercial assets to speed up their development.

## What solution would you like?

Support FBX file import through a loader/processor.

A loader like [bevy_mod_fbx](https://github.com/nicopap/bevy_mod_fbx) could be upstreamed.

Alternatively, [like Godot](https://github.com/godotengine/godot/pull/59653), leverage [fbx2gltf](https://github.com/godotengine/FBX2glTF) or [ufbx](https://github.com/ufbx/ufbx)(current Godot default) within the engine to perform the conversion seamlessly for users

## What alternative(s) have you considered?

Relying on 3rd party crates (none seem up to date), or manually converting via Blender.

Add fbx -> gltf conversion into the bevy cli tool.

## Additional context

I searched for "fbx" issues previously, but couldn't find one in the repo. I assume this has been thought of before, but I wanted to formalise it.

**Comment by mrchantey** [comment]
Great idea centralizing discussion on additional 3d model formats. I think it makes sense to have a clear path for how to do this in bevy, and for efforts to be consolidated.

I guess conversion to gltf is the discussion here, deep conversionless integration would add quite a bit of complexity both in implementation and maintenance.

Seeing as godot has picked up the reigns of [FBX2glTF](https://github.com/godotengine/FBX2glTF), us not having our own one might be more of a 'feature than a bug', as in being able to share maintenance of engine agnostic tools allows both communities to focus on other things.

Regarding batch conversion I'd personally prefer to use a cli with a simple recursive script instead of something deeply integrated and automatic, particularly because the process is lossy, but of course thats easy to say for people who write scripts and use cli tools every day. 

I suppose the exact workflow for non-technical people might be easier to visualize once we have an editor, like some kind of extension that runs FBX2glTF whenever an FBX file is dragged and dropped into the asset section.

**Comment by tbillington** [comment]
> I guess conversion to gltf is the discussion here

Yes. Bevy seems quite gltf-centric so converting to gltf is probably the most feasible option that also avoids increasing the logic in bevy "core".

> I suppose the exact workflow for non-technical people might be easier to visualize once we have an editor, like some kind of extension that runs FBX2glTF whenever an FBX file is dragged and dropped into the asset section.

This is what Godot does, and is likely like the most user friendly solution, especially if you include teams with people who may not be CLI fluent. In my opinion, the original file should still be kept in `assets`, but the resulting gltf would end up in the directory from `asset_processor`.

**Comment by VitalyAnkh** [comment]
> Seeing as godot has picked up the reigns of [FBX2glTF](https://github.com/godotengine/FBX2glTF)

Godot 4.3 introduced a [native FBX importer, ufbx](https://godotengine.org/article/introducing-the-improved-ufbx-importer-in-godot-4-3/), and enabled it by default. This change is more user-friendly and greatly improves the overall experience.

Since glTF and FBX are inherently different formats, direct conversion often results in information inconsistencies and other undesirable outcomes. The best solution for FBX support is to implement a native importer. Upstreaming [bevy_mod_fbx](https://github.com/nicopap/bevy_mod_fbx) is a positive step.

**Comment by tbillington** [comment]
@VitalyAnkh I agree that native support would be the best version of support. But my thoughts are that the fbx -> gltf route is much more palatable for the Bevy project at this point in time.

`bevy_mod_fbx` does a great job, but the underlying library [`fbxcel-dom`](https://github.com/lo48576/fbxcel-dom)/[`fbxcell`](https://github.com/lo48576/fbxcel) is only maintained by one person, and together with `bevy_mod_fbx` there is quite a lot of work left to be done before it's anywhere close to Bevy's gltf support.

<details><summary>Unimplemented features & limitations of `bevy_mod_fbx`</summary>
<p>

Planned features

    Skeleton rig imports
    bevy_animation support
    Optional lambert material shader support
    Optional phong shading model support
    Extended compatibility:
        IndexToDirect
        Handle file-based axis properties
        Handle backed cameras & lights
        N-gon triangulation

Limitations

    FBX v7.4 & 7.5 are the only supported versions
    FBX doesn't support multiple scenes in single file, use multiple files instead
    There are no plans for loading ASCII format, export FBX as binary v7.4/7.5
    There is no support for complex shapes at the moment, see https://github.com/HeavyRain266/bevy_mod_fbx/issues/11

</p>
</details> 

So without someone or a group to commit to improving the situation with `bevy_mod_fbx`, it doesn't appear the right choice at this point in time, in my opinion.

**Comment by JulianKnodt** [comment]
I'm currently working on a parser for FBX/GLB/OBJ/PLY [here](https://github.com/JulianKnodt/pars3d), what features would be necessary to incorporate it into bevy?

**Comment by tbillington** [comment]
Seems like Blender contributors are prototyping integrating [ufbx](https://github.com/ufbx/ufbx) too. [Issue](https://projects.blender.org/blender/blender/issues/131304) and [PR](https://projects.blender.org/blender/blender/pulls/132406).

I just saw that ufbx is a single source file C99/C++11 library, so it's possible generating Rust bindings might not be that difficult.

**Comment by FizzWizZleDazzle** [comment]
Is anyone working on generating Rust bindings?

**Comment by VitalyAnkh** [comment]
ufbx has an official rust binding: https://github.com/ufbx/ufbx-rust.

**Comment by FizzWizZleDazzle** [comment]
Yeah I just saw that.

**Comment by VitalyAnkh** [comment]
I have submitted a PR to add FBX loader based on ufbx: https://github.com/bevyengine/bevy/pull/19534.

**Comment by Igor-dvr** [comment]
It seems FBX is going to be deprecated and tools moving to OpenUSD/GLTF
https://www.youtube.com/watch?v=_quZ1QRysSU&t=1379s

**Comment by FizzWizZleDazzle** [comment]
> It seems FBX is going to be deprecated and tools moving to OpenUSD/GLTF https://www.youtube.com/watch?v=_quZ1QRysSU&t=1379s

Even if FBX gets deprecated, there’s still a huge amount of existing assets and pipelines built around it. Many libraries/marketplaces still distribute FBX, and it often carries data like rigs, animation curves, and material setups that aren’t always preserved in glTF/USD.

**Comment by Igor-dvr** [comment]
> and it often carries data like rigs, animation curves, and material setups that aren’t always preserved in glTF/USD.

It seems there are also plans to implement FBX features in GLTF/USD.

![Image](https://github.com/user-attachments/assets/079ece37-6c32-4e1b-83c5-dae80c74ab1d)

**Comment by Igor-dvr** [comment]
And I'm not against this. Just shared the info.
FBX/OpenUSD curve support would be nice. Gltf doesn't support curves.

**Comment by Igor-dvr** [comment]
> A Bevy plugin for loading FBX files using the ufbx library

https://github.com/FizzWizZleDazzle/bevy_ufbx
